compiler/rustc_query_system/src/dep_graph/graph.rs

```diff
@@ -36,6 +43,9 @@ pub struct DepGraph<K: DepKind> {
     virtual_dep_node_index: Lrc<AtomicU32>,
 }
 
+#[cfg(parallel_compiler)]
+unsafe impl<K: DepKind> Sync for DepGraph<K> {}
```

Comment:
This feels very hard to justify and hard to maintain -- it at minimum needs a comment on it, but I think we really ought to find a way to avoid unsafe impls on "random" data structures in the compiler, it doesn't scale to maintain those as we iterate on the fields here.

Reply:
Yes, these data structures should be encapsulated using a unified data structure. We should have a way to ensure that these data structures are not used in a multithreaded environment when they are in a "non-shared" state. I think `fragile` and `WorkerLocal` can be referenced here.